### DISCUSSION TITLE
How to explicitly skip checking part of a schema in zod

### DISCUSSION THREAD

**Discussion by doberkofler** [colinhacks/zod#1575] [Q&A] [answered]
I would like to understand if and how it is possible to skip validating parts of a schema in zod?

In the following example, I would like to validate the foo schema to make sure that the object contains a property id of type number and a property data of type array but (maybe cause there is a lot of data) I would like to prevent validating all the actual array entries in data.

```typescript
import {z} from 'zod';

const foo = z.object({
   id: z.number(),
   data: z.array(z.string()),
});
```

**Comment by scotttrinh** [comment, marked as answer]
```ts
const foo = z.object({
   id: z.number(),
   data: z.array(z.any()),
});
```

This would do it, but your type will be:
```ts
type Foo = { id: string; data: any[] };
```

If you truly don't care about the data (maybe you're just passing it through to some other service?) you can use `z.unknown` like:

```ts
const foo = z.object({
   id: z.number(),
   data: z.array(z.unknown()),
});
// { id: string; data: unknown[] }
```

If you want to have a type for the data array values, but just skip checking them, you can make a custom type that just doesn't do any runtime checking, but you're explicitly opting to have unchecked data here, so I would be very very careful:

```ts
const dataItem = z.custom<DataItem>(); // type DataItem defined by you elsewhere
const foo = z.object({
   id: z.number(),
   data: z.array(dataItem),
});
// { id: string; data: DataItem[] }
```

**Reply by doberkofler** [reply]
Thank you for the quick feedback!

In my typical use cases the first two options would not be "correct" because I will also use the inferred TypeScript type and it should be fully qualified: `{id: number, data: string[]}`

The third option looks very promising. Did I completely overlook the `z.custom` type definition in the documentation or is this a hidden feature? Did I understand you correctly, and `z.custom` would allow me to specify a TypeScript type that will be inferred and the final type definition but not validated or converted at all?

Thank you!

**Comment by JacobWeisenburger** [comment]
I suggest using [`.pick` or `.omit`](https://github.com/colinhacks/zod#pickomit)

```ts
const foo = z.object( {
    id: z.number(),
    data: z.array( z.string() ),
} )
console.log( foo.safeParse( { id: 1 } ).success ) // false
console.log( foo.pick( { id: true } ).safeParse( { id: 1 } ).success ) // true
console.log( foo.omit( { data: true } ).safeParse( { id: 1 } ).success ) // true
```
